### PR TITLE
Refactor transport layer

### PR DESCRIPTION
**Pull request by jean-roland**
Transport layer refactor to improve maintainability and modularity in prevision of adding a new transport.

Instead of compile-time specific code removal spread everywhere, files call functions that are swapped to dummy versions when the feature is removed. 

This may increase slightly code size when removing the feature but makes it much easier to expand the number of transports.

+ added a few "early return" pattern here and there

### CODE REVIEW

**Review by cguimaraes — COMMENTED** [review]
If possible avoid the early returns, unless they are really needed.
If you still keep them, document them (including the reason).

File: include/zenoh-pico/transport/multicast.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+#ifndef ZENOH_PICO_MULTICAST_H
+#define ZENOH_PICO_MULTICAST_H
+
+#include "zenoh-pico/api/types.h"
+
+bool _zp_is_multicast_here(void);
```
Comment (Mallets) [inline include/zenoh-pico/transport/multicast.h:20]: What's the purpose of this function?

File: include/zenoh-pico/transport/unicast.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+#ifndef ZENOH_PICO_UNICAST_H
+#define ZENOH_PICO_UNICAST_H
+
+#include "zenoh-pico/api/types.h"
+
+bool _zp_is_unicast_here(void);
```
Comment (Mallets) [inline include/zenoh-pico/transport/unicast.h:20]: What's the purpose of this function?

File: include/zenoh-pico/transport/multicast/lease.h [hunk 3]
```diff
@@ -12,17 +12,14 @@
...
 #include "zenoh-pico/transport/transport.h"
 
-int8_t _z_send_keep_alive(_z_transport_t *zt);
-int8_t _zp_unicast_send_keep_alive(_z_transport_unicast_t *ztu);
 int8_t _zp_multicast_send_keep_alive(_z_transport_multicast_t *ztm);
-
-void *_zp_lease_task(void *zt_arg);             // The argument is void* to avoid incompatible pointer types in tasks
-void *_zp_unicast_lease_task(void *ztu_arg);    // The argument is void* to avoid incompatible pointer types in tasks
+int _zp_multicast_start_lease_task(_z_transport_t *zt, _z_task_attr_t *attr, _z_task_t *task);
+int _zp_multicast_stop_lease_task(_z_transport_t *zt);
```
Comment (cguimaraes) [inline include/zenoh-pico/transport/multicast/lease.h:22]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: include/zenoh-pico/transport/multicast/lease.h [hunk 4]
```diff
@@ -12,17 +12,14 @@
...
 
 #include "zenoh-pico/transport/transport.h"
 
-int8_t _z_send_keep_alive(_z_transport_t *zt);
-int8_t _zp_unicast_send_keep_alive(_z_transport_unicast_t *ztu);
 int8_t _zp_multicast_send_keep_alive(_z_transport_multicast_t *ztm);
-
-void *_zp_lease_task(void *zt_arg);             // The argument is void* to avoid incompatible pointer types in tasks
-void *_zp_unicast_lease_task(void *ztu_arg);    // The argument is void* to avoid incompatible pointer types in tasks
+int _zp_multicast_start_lease_task(_z_transport_t *zt, _z_task_attr_t *attr, _z_task_t *task);
```
Comment (cguimaraes) [inline include/zenoh-pico/transport/multicast/lease.h:21]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: include/zenoh-pico/transport/multicast/read.h [hunk 5]
```diff
@@ -12,17 +12,14 @@
...
 
 #include "zenoh-pico/transport/transport.h"
 
-int8_t _z_read(_z_transport_t *zt);
-int8_t _zp_unicast_read(_z_transport_unicast_t *ztu);
 int8_t _zp_multicast_read(_z_transport_multicast_t *ztm);
-
-void *_zp_read_task(void *zt_arg);             // The argument is void* to avoid incompatible pointer types in tasks
-void *_zp_unicast_read_task(void *ztu_arg);    // The argument is void* to avoid incompatible pointer types in tasks
+int _zp_multicast_start_read_task(_z_transport_t *zt, _z_task_attr_t *attr, _z_task_t *task);
```
Comment (cguimaraes) [inline include/zenoh-pico/transport/multicast/read.h:21]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: include/zenoh-pico/transport/multicast/read.h [hunk 6]
```diff
@@ -12,17 +12,14 @@
...
+int _zp_multicast_stop_read_task(_z_transport_t *zt);
```
Comment (cguimaraes) [inline include/zenoh-pico/transport/multicast/read.h:22]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: include/zenoh-pico/transport/unicast/lease.h [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+//
+
+#ifndef ZENOH_PICO_UNICAST_LEASE_H
+#define ZENOH_PICO_UNICAST_LEASE_H
+
+#include "zenoh-pico/transport/transport.h"
+
+int8_t _zp_unicast_send_keep_alive(_z_transport_unicast_t *ztu);
+int _zp_unicast_start_lease_task(_z_transport_t *zt, _z_task_attr_t *attr, _z_task_t *task);
+int _zp_unicast_stop_lease_task(_z_transport_t *zt);
```
Comment (cguimaraes) [inline include/zenoh-pico/transport/unicast/lease.h:22]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: include/zenoh-pico/transport/unicast/lease.h [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+#ifndef ZENOH_PICO_UNICAST_LEASE_H
+#define ZENOH_PICO_UNICAST_LEASE_H
+
+#include "zenoh-pico/transport/transport.h"
+
+int8_t _zp_unicast_send_keep_alive(_z_transport_unicast_t *ztu);
+int _zp_unicast_start_lease_task(_z_transport_t *zt, _z_task_attr_t *attr, _z_task_t *task);
```
Comment (cguimaraes) [inline include/zenoh-pico/transport/unicast/lease.h:21]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: include/zenoh-pico/transport/unicast/read.h [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+#ifndef ZENOH_PICO_UNICAST_READ_H
+#define ZENOH_PICO_UNICAST_READ_H
+
+#include "zenoh-pico/transport/transport.h"
+
+int8_t _zp_unicast_read(_z_transport_unicast_t *ztu);
+int _zp_unicast_start_read_task(_z_transport_t *zt, _z_task_attr_t *attr, _z_task_t *task);
```
Comment (cguimaraes) [inline include/zenoh-pico/transport/unicast/read.h:21]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: include/zenoh-pico/transport/unicast/read.h [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+int _zp_unicast_stop_read_task(_z_transport_t *zt);
```
Comment (cguimaraes) [inline include/zenoh-pico/transport/unicast/read.h:22]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: src/transport/multicast/lease.c [hunk 11]
```diff
@@ -67,6 +69,25 @@ int8_t _zp_multicast_send_keep_alive(_z_transport_multicast_t *ztm) {
     return ret;
 }
 
+int _zp_multicast_start_lease_task(_z_transport_t *zt, _z_task_attr_t *attr, _z_task_t *task) {
```
Comment (cguimaraes) [inline src/transport/multicast/lease.c:72]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: src/transport/multicast/lease.c [hunk 12]
```diff
@@ -67,6 +69,25 @@ int8_t _zp_multicast_send_keep_alive(_z_transport_multicast_t *ztm) {
...
+    zt->_transport._multicast._lease_task_running = true;
+    // Init task
+    if (_z_task_init(task, attr, _zp_multicast_lease_task, &zt->_transport._multicast) != _Z_RES_OK) {
+        zt->_transport._multicast._lease_task_running = false;
+        return _Z_ERR_SYSTEM_TASK_FAILED;
+    }
+    return _Z_RES_OK;
+}
+
+int _zp_multicast_stop_lease_task(_z_transport_t *zt) {
```
Comment (cguimaraes) [inline src/transport/multicast/lease.c:86]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: src/transport/multicast/lease.c [hunk 13]
```diff
@@ -162,5 +183,26 @@ void *_zp_multicast_lease_task(void *ztm_arg) {
 
     return 0;
 }
+#else
+int8_t _zp_multicast_send_keep_alive(_z_transport_multicast_t *ztm) {
+    _ZP_UNUSED(ztm);
+    return _Z_ERR_TRANSPORT_NOT_AVAILABLE;
+}
 
+int _zp_multicast_start_lease_task(_z_transport_t *zt, _z_task_attr_t *attr, _z_task_t *task) {
```
Comment (cguimaraes) [inline src/transport/multicast/lease.c:192]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: src/transport/multicast/lease.c [hunk 14]
```diff
@@ -162,5 +183,26 @@ void *_zp_multicast_lease_task(void *ztm_arg) {
...
+    _ZP_UNUSED(zt);
+    _ZP_UNUSED(attr);
+    _ZP_UNUSED(task);
+    return _Z_ERR_TRANSPORT_NOT_AVAILABLE;
+}
+
+int _zp_multicast_stop_lease_task(_z_transport_t *zt) {
```
Comment (cguimaraes) [inline src/transport/multicast/lease.c:199]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: src/transport/multicast/read.c [hunk 15]
```diff
@@ -38,6 +39,25 @@ int8_t _zp_multicast_read(_z_transport_multicast_t *ztm) {
     return ret;
 }
 
+int _zp_multicast_start_read_task(_z_transport_t *zt, _z_task_attr_t *attr, _z_task_t *task) {
```
Comment (cguimaraes) [inline src/transport/multicast/read.c:42]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: src/transport/multicast/read.c [hunk 16]
```diff
@@ -38,6 +39,25 @@ int8_t _zp_multicast_read(_z_transport_multicast_t *ztm) {
...
+    zt->_transport._multicast._read_task_running = true;
+    // Init task
+    if (_z_task_init(task, attr, _zp_multicast_read_task, &zt->_transport._multicast) != _Z_RES_OK) {
+        zt->_transport._multicast._read_task_running = false;
+        return _Z_ERR_SYSTEM_TASK_FAILED;
+    }
+    return _Z_RES_OK;
+}
+
+int _zp_multicast_stop_read_task(_z_transport_t *zt) {
```
Comment (cguimaraes) [inline src/transport/multicast/read.c:56]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: src/transport/multicast/read.c [hunk 17]
```diff
@@ -117,5 +137,26 @@ void *_zp_multicast_read_task(void *ztm_arg) {
 
     return NULL;
 }
+#else
+int8_t _zp_multicast_read(_z_transport_multicast_t *ztm) {
+    _ZP_UNUSED(ztm);
+    return _Z_ERR_TRANSPORT_NOT_AVAILABLE;
+}
 
+int _zp_multicast_start_read_task(_z_transport_t *zt, _z_task_attr_t *attr, _z_task_t *task) {
```
Comment (cguimaraes) [inline src/transport/multicast/read.c:146]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: src/transport/multicast/read.c [hunk 18]
```diff
@@ -117,5 +137,26 @@ void *_zp_multicast_read_task(void *ztm_arg) {
...
+    _ZP_UNUSED(zt);
+    _ZP_UNUSED(attr);
+    _ZP_UNUSED(task);
+    return _Z_ERR_TRANSPORT_NOT_AVAILABLE;
+}
+
+int _zp_multicast_stop_read_task(_z_transport_t *zt) {
```
Comment (cguimaraes) [inline src/transport/multicast/read.c:153]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: src/transport/unicast/lease.c [hunk 19]
```diff
@@ -28,6 +29,25 @@ int8_t _zp_unicast_send_keep_alive(_z_transport_unicast_t *ztu) {
     return ret;
 }
 
+int _zp_unicast_start_lease_task(_z_transport_t *zt, _z_task_attr_t *attr, _z_task_t *task) {
```
Comment (cguimaraes) [inline src/transport/unicast/lease.c:32]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: src/transport/unicast/lease.c [hunk 20]
```diff
@@ -28,6 +29,25 @@ int8_t _zp_unicast_send_keep_alive(_z_transport_unicast_t *ztu) {
...
+    zt->_transport._unicast._lease_task_running = true;
+    // Init task
+    if (_z_task_init(task, attr, _zp_unicast_lease_task, &zt->_transport._unicast) != _Z_RES_OK) {
+        zt->_transport._unicast._lease_task_running = false;
+        return _Z_ERR_SYSTEM_TASK_FAILED;
+    }
+    return _Z_RES_OK;
+}
+
+int _zp_unicast_stop_lease_task(_z_transport_t *zt) {
```
Comment (cguimaraes) [inline src/transport/unicast/lease.c:46]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: src/transport/unicast/lease.c [hunk 21]
```diff
@@ -87,5 +107,26 @@ void *_zp_unicast_lease_task(void *ztu_arg) {
 
     return 0;
 }
+#else
+int8_t _zp_unicast_send_keep_alive(_z_transport_unicast_t *ztu) {
+    _ZP_UNUSED(ztu);
+    return _Z_ERR_TRANSPORT_NOT_AVAILABLE;
+}
 
+int _zp_unicast_start_lease_task(_z_transport_t *zt, _z_task_attr_t *attr, _z_task_t *task) {
```
Comment (cguimaraes) [inline src/transport/unicast/lease.c:116]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: src/transport/unicast/lease.c [hunk 22]
```diff
@@ -87,5 +107,26 @@ void *_zp_unicast_lease_task(void *ztu_arg) {
...
+    _ZP_UNUSED(zt);
+    _ZP_UNUSED(attr);
+    _ZP_UNUSED(task);
+    return _Z_ERR_TRANSPORT_NOT_AVAILABLE;
+}
+
+int _zp_unicast_stop_lease_task(_z_transport_t *zt) {
```
Comment (cguimaraes) [inline src/transport/unicast/lease.c:123]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: src/transport/unicast/read.c [hunk 23]
```diff
@@ -36,6 +36,25 @@ int8_t _zp_unicast_read(_z_transport_unicast_t *ztu) {
     return ret;
 }
 
+int _zp_unicast_start_read_task(_z_transport_t *zt, _z_task_attr_t *attr, _z_task_t *task) {
```
Comment (cguimaraes) [inline src/transport/unicast/read.c:39]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: src/transport/unicast/read.c [hunk 24]
```diff
@@ -36,6 +36,25 @@ int8_t _zp_unicast_read(_z_transport_unicast_t *ztu) {
...
+    zt->_transport._unicast._read_task_running = true;
+    // Init task
+    if (_z_task_init(task, attr, _zp_unicast_read_task, &zt->_transport._unicast) != _Z_RES_OK) {
+        zt->_transport._unicast._read_task_running = false;
+        return _Z_ERR_SYSTEM_TASK_FAILED;
+    }
+    return _Z_RES_OK;
+}
+
+int _zp_unicast_stop_read_task(_z_transport_t *zt) {
```
Comment (cguimaraes) [inline src/transport/unicast/read.c:53]: As convention, return `int8_t` instead.
This goes back to a previous discussion on having a list of return codes.

File: include/zenoh-pico/transport/unicast.h (hunk 2 above)
Comment (jean-roland) [reply include/zenoh-pico/transport/unicast.h:20]: This one is unused, just made a mirror of the multicast one

File: include/zenoh-pico/transport/multicast.h (hunk 1 above)
Comment (jean-roland) [reply include/zenoh-pico/transport/multicast.h:20]: It's used on the scouting function it replaces a `#if` that was doing the same.

File: src/session/scout.c [hunk 25]
```diff
@@ -140,14 +141,9 @@ _z_hello_list_t *_z_scout_inner(const z_what_t what, _z_id_t zid, const char *lo
...
     // Scout on multicast
-#if Z_FEATURE_MULTICAST_TRANSPORT == 1
-    ret = __z_scout_loop(&wbf, locator, timeout, exit_on_first);
-#else
-    (void)(locator);
-    (void)(timeout);
-    (void)(exit_on_first);
-#endif  // Z_FEATURE_MULTICAST_TRANSPORT == 1
-
+    if (_zp_is_multicast_here()) {
```
Comment (jean-roland) [inline src/session/scout.c:144]: @Mallets this is the only place where the `is_multicast_here` function is used

Comment (Mallets) [reply src/session/scout.c:144]: Looking at the code, it seems that scouting is only performed when *multicast transport* is activated.
However, https://github.com/jean-roland/zenoh-pico/blob/20640afe4d8f442272597b56604e514f84a0bb75/src/session/scout.c#L36 already provides the necessary check on UDP multicast support. Therefore, I believe that `_zp_is_multicast_here()` is redundant and could be removed.

**Review by Mallets — APPROVED** [review]

**Pull request merged**